server/aws-lsp-codewhisperer/src/language-server/localProjectContext/localProjectContextController.ts

```diff
@@ -0,0 +1,204 @@
+
+export class LocalProjectContextController {
+    private static instance: LocalProjectContextController | undefined
+
+    private workspaceFolders: WorkspaceFolder[]
+    private _vecLib?: VectorLibAPI
+    private readonly fileExtensions: string[]
+    private readonly clientName: string
+    private readonly log: Logging
+
+    constructor(clientName: string, workspaceFolders: WorkspaceFolder[], logging: Logging) {
+        this.fileExtensions = Object.keys(languageByExtension)
+        this.workspaceFolders = workspaceFolders
+        this.clientName = clientName
+        this.log = logging
+    }
+
+    public static getInstance() {
+        if (!this.instance) {
+            throw new Error('LocalProjectContextController not initialized')
+        }
+        return this.instance
+    }
+
+    public async init(vectorLib?: any): Promise<void> {
+        try {
+            const vecLib = vectorLib ?? (await import(path.join(LIBRARY_DIR, 'dist', 'extension.js')))
+            const root = this.findCommonWorkspaceRoot(this.workspaceFolders)
+            this._vecLib = await vecLib.start(LIBRARY_DIR, this.clientName, root)
+            await this.buildIndex()
+            LocalProjectContextController.instance = this
+        } catch (error) {
+            this.log.error('Vector library failed to initialize:' + error)
+        }
+    }
+
+    public async dispose(): Promise<void> {
+        if (this._vecLib) {
+            await this._vecLib?.clear?.()
+            this._vecLib = undefined
+        }
+    }
+
+    public async updateIndex(filePaths: string[], operation: UpdateMode): Promise<void> {
+        if (!this._vecLib) {
+            return
+        }
+
+        try {
+            await this._vecLib?.updateIndexV2(filePaths, operation)
+        } catch (error) {
+            this.log.error(`Error updating index: ${error}`)
+        }
+    }
+
+    private async buildIndex(): Promise<void> {
+        try {
+            if (this._vecLib) {
+                const sourceFiles = await this.processWorkspaceFolders(this.workspaceFolders)
+                const rootDir = this.findCommonWorkspaceRoot(this.workspaceFolders)
+                await this._vecLib?.buildIndex(sourceFiles, rootDir, 'all')
+            }
+        } catch (error) {
+            this.log.error(`Error building index: ${error}`)
+        }
+    }
+
+    public async updateWorkspaceFolders(added: WorkspaceFolder[], removed: WorkspaceFolder[]): Promise<void> {
+        try {
+            const afterRemovals = this.workspaceFolders.filter(
+                existing => !removed.some(removal => this.areWorkspaceFoldersEqual(existing, removal))
+            )
+
+            const merged = [...afterRemovals]
+            for (const addition of added) {
+                if (!merged.some(existing => this.areWorkspaceFoldersEqual(existing, addition))) {
+                    merged.push(addition)
+                }
+            }
+            if (this._vecLib) {
```

Comment:
Shouldn't we check for this on the top of this method - I do not think any code needs to be executed if library is disabled.

Reply:
In general, you could introduce NoOp controller for cases when indexing is not supported - it would be more readable and less error-prone. Or even check for the `enabled` flag earlier - in server.